kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/data/StringTableTest.kt

```diff
@@ -0,0 +1,213 @@
+         row(1, "bad | good", "name")
+      )
+   }
+
+   val resourcesDir = File("src/jvmTest/resources/table")
+
+   test("happy path for reading a table from a file") {
+      val file = resourcesDir.resolve("users-valid.table")
+      table(headers, file, transform) shouldBe expectedTable
+   }
+
+   test("Validating table files") {
+
+      shouldThrowMessage("Can't read table file") {
+         val file = resourcesDir.resolve("users-does-not-exist.table")
+         table(headers, file, transform)
+      }
+
+      shouldThrowMessage("Table file must have a .table extension") {
+         val file = resourcesDir.resolve("users-invalid-extension.csv")
+         table(headers, file, transform)
+      }
+
+      shouldThrowMessage("Table file must have a header") {
+         val file = resourcesDir.resolve("users-invalid-empty.table")
+         table(headers, file, transform)
+      }
+
+      shouldThrowMessage(
+         """
+         Missing elements from index 2
+         expected:<["id", "username", "fullName"]> but was:<["id", "username"]>
+         """.trimIndent()
+      ) {
+         val file = resourcesDir.resolve("users-invalid-header.table")
+         table(headers, file, transform)
+      }
+   }
+
+   data class UserInfo(val username: String, val fullName: String)
+   val usersTable = table(
+      headers("id", "UserInfo"),
+      row(4, UserInfo("jmfayard", "Jean-Michel Fayard")),
+      row(6, UserInfo("louis", "Louis Caugnault"))
+   )
+
+   context("file.writeTable - success") {
+      val expectedFileContent = """
+id | username | fullName
+4  | jmfayard | Jean-Michel Fayard
+6  | louis    | Louis Caugnault
+      """.trim()
+
+      test("happy path") {
+         val file = tempfile(suffix = ".table")
+         val rows = usersTable.mapRows { (id, userInfo) ->
+            row(id.toString(), userInfo.username, userInfo.fullName)
+         }
+         val fileContent = file.writeTable(headers("id", "username", "fullName"), rows)
+         file.readText() shouldBe expectedFileContent
+         fileContent shouldBe expectedFileContent
+      }
+
+      test("columns should be aligned") {
+         fun row(i: Int): Row3<String, String, String> {
+            val value = "$i".repeat(i)
+            return row(value, value, value)
+         }
+
+         val table = table(
+            headers("a", "b", "c"),
+            row(2),
+            row(4),
+            row(6),
+         )
+         tempfile(suffix = ".table").writeTable(table.headers, table.rows) shouldBe """
+a      | b      | c
+22     | 22     | 22
+4444   | 4444   | 4444
+666666 | 666666 | 666666
+         """.trimIndent()
+      }
+
+      test("| should be escaped") {
+         val table = mapOf("greeting" to "Hello || world").toTable()
+         val file = tempfile(suffix = ".table")
+         file.writeTable(table.headers, table.rows) shouldBe """
+            key      | value
+            greeting | Hello \|\| world
+         """.trimIndent()
+      }
+   }
```

Comment:
I think we also need to add escaping of backslash itself. E.g.: 

```
Hello \|\| World ->  (Hello || World)  // 1 item
Hello \\|| World -> (Hello \,, World)   // 3 items
Hello \\\|\\\| World -> (Hello \|\| World) // 1 item
```

Reply:
done